feat/select-messages and download image

MasterK0927 opened the pull request:
**Resolves #115**  
**Resolves #133**

1. **Message Selection**: Users can now select a message by long pressing on mobile devices (using the `svelte-gesture` `on:press` event) or by clicking on desktop.  
   
2. **Action Toolbar**: Upon selecting a message, a toolbar appears beneath the message with options based on the message type. The options include:
   - **Cancel** (always visible)
   - **Copy** (for text-based messages)
   - **Download** (for image-based messages)
   - The edge case where a message contains both text and an image is handled correctly, showing both the **Copy** and **Download** buttons.

3. **Single Selection**: Only one message can be selected at a time. If the user clicks anywhere outside the selected message, the message is deselected.

4. **Download Support**: The **Download** button uses Tauri's dialog plugin to allow users to download images attached to messages.

6. **Strings Added**: Added new strings in the conversations.json
   - [`Line 11`](https://github.com/holochain-apps/volla-messages/commit/562f7f309ddceff894b9ab6f72ce1f200a56e104#diff-f2e3c1233905291d98a6ce810906f8f7fe1874222ae4bb86d3b3d62d5156622aR11)
   - [`Line 34`](https://github.com/holochain-apps/volla-messages/commit/562f7f309ddceff894b9ab6f72ce1f200a56e104#diff-f2e3c1233905291d98a6ce810906f8f7fe1874222ae4bb86d3b3d62d5156622aR34)

MasterK0927 committed:
option for downloading images

MasterK0927 committed:
clustered the download image and seleted messages together

MasterK0927 committed:
refactor code to address existing comments, removing writable logic implementation.

MasterK0927 committed:
Refactor to remove writable logic for selectedMessages, allowing Svelte to handle reactivity.

MasterK0927 committed:
fix press gesture to select on mobile devices

mattyg commented on ui/src/types.ts:
```diff
@@ -12,40 +12,40 @@ import type {
   DeleteLink,
   MembraneProof,
   ClonedCell,
-} from "@holochain/client";
```
It looks like there are still formatting changes from conflicting formatter setups. Were you able to install the recommended plugins in vscode?

MasterK0927 replied:
Fixed the formatting mismatch

mattyg commented on ui/src/routes/conversations/[id]/+page.svelte:
```diff
@@ -80,6 +83,38 @@
     })
   }
 
+  const downloadImage = async (image: Image) => {
```
Please move this function into the new MessageActions component, since that is the only place it is used.

MasterK0927 replied:
Moved the download utility function inside `MessageActions` component.

mattyg commented on ui/src/routes/conversations/[id]/+page.svelte:
```diff
@@ -236,6 +271,51 @@
+
+  let selectedMessageHash: string | null = null;
+    
+    function toggleMessageSelection(messageHash: string) {
```
Please run the formatting so these have consistent indentation

MasterK0927 replied:
Fixed the formatting issues.

mattyg commented on ui/src/routes/conversations/[id]/+page.svelte:
```diff
@@ -236,6 +271,51 @@
+      }
+    }
+  
+    function handlePress(messageHash: string, event: PressCustomEvent) {
```
```suggestion
    function handleMessagePress(messageHash: string, event: PressCustomEvent) {
```

good to have consistent naming for clarity

MasterK0927 replied:
Fixed the inconsistent naming issue.

mattyg commented on ui/src/routes/conversations/[id]/+page.svelte:
```diff
@@ -236,6 +271,51 @@
+  
+    function handleMessageClick(messageHash: string, event: MouseEvent) {
+      if(!isMobile()){
+        event.preventDefault();
```
Please use svelte's event modifiers rather than passing the event up. Here's an example:

```javascript
<button on:click|preventDefault={() => alert("hello")}>My Button</button>
```

Here's the svelte docs about it: https://v4.svelte.dev/docs/element-directives#on-eventname

MasterK0927 replied:
Updated the code to use Svelte's event modifiers

mattyg commented on ui/src/routes/conversations/[id]/+page.svelte:
```diff
@@ -339,31 +419,52 @@
-                    <span class='min-w-6 inline-block'></span>
+              <li class='mt-auto {!message.hideDetails && "mt-3"} relative {isSelected ? 'selected-message-container' : ''}'>
+                <!-- svelte-ignore a11y-click-events-have-key-events -->
+                <!-- svelte-ignore a11y-no-static-element-interactions -->
```
Can we fix these rather than ignoring them?

MasterK0927 replied:
Fixed by changing `div` to `button`.

mattyg commented on ui/src/routes/conversations/[id]/+page.svelte:
```diff
@@ -427,4 +539,23 @@
+    border-top-right-radius: 0;
+    border-bottom-left-radius: 0.75rem;
+    border-bottom-right-radius: 0.75rem;
+  }
```
Please use tailwind classes for styling if possible (I know there are a few places in the codebase right now that are not using tailwind and I'd like to remove those).

MasterK0927 replied:
Migrated all the vanilla css styling to tailwind.

mattyg commented on ui/src/lib/MessageActions.svelte:
```diff
@@ -0,0 +1,86 @@
+    import { copyToClipboard } from '$lib/utils';
+    
+    export let message: Message;
+    export let messageHash: string;
```
The `Message` type has a field `hash`, so we don't need another prop to pass in the messageHash.

MasterK0927 replied:
Removed the redundant messageHash prop.

mattyg commented on ui/src/lib/MessageActions.svelte:
```diff
@@ -0,0 +1,86 @@
+    export let downloadImage: (image: any) => Promise<void>;
+    export let unselectMessage: () => void;
+  
+    const messageActions: MessageAction[] = [
```
Lets move this directly in the svelte template, rather than defining the logic here looping through each action. With that approach we'll have 3  functions: `copy`, `cancel` and `download`. And 2 derived values: `hasText` and `hasImages`. 

This follows the common conventions for svelte and because of that will be clearer.

MasterK0927 replied:
Logic moved to the Svelte template with copy, cancel, download functions and hasText, hasImages derived values for clarity and convention.

mattyg reviewed (changes requested):
Thanks! I made a couple comments inline and have a few more change requests here.

- Thanks for naming this branch clearly! In the future, please keep all revisions made after a review in the same branch. It makes it much easier to confirm that any feedback or issues have been addressed.

- The "Download" button icon is a different color than the others.

- Lets just remove the "cancel" button and rely on clicking outside the message to deselect it. I think that's simpler UX.

- Please make all the buttons slightly smaller -- roughly 2/3 their current size.

- The buttons overflow the container in small viewports (see the screenshot below). They may need to be even smaller for small viewports and/or wrap around. You can use the tailwindcss modifiers to adjust styling based on viewport size.
![image](https://github.com/user-attachments/assets/c70f2561-c486-40fa-af06-8666b6a1b5af)

- The "copy" icon seems to scale down in smaller viewports, while the others don't. All of them should be styled consistently.

- When message A is selected, then message B is clicked, lets have it only deselect Message A, and require the user to click message B again to select it. I think that UX feels more similar to other apps.

- If you import a new js package, please add it to package.json. It looks like `@tauri-apps/plugin-dialog` was not added. (I went ahead and added it and pushed to this branch)

MasterK0927 committed:
fixed formatting issues and svelte based changes

MasterK0927 commented:
> Thanks! I made a couple comments inline and have a few more change requests here.
> 
> - Thanks for naming this branch clearly! In the future, please keep all revisions made after a review in the same branch. It makes it much easier to confirm that any feedback or issues have been addressed.
> 
> - The "Download" button icon is a different color than the others.
> 
> - Lets just remove the "cancel" button and rely on clicking outside the message to deselect it. I think that's simpler UX.
> 
> - Please make all the buttons slightly smaller -- roughly 2/3 their current size.
> 
> - The buttons overflow the container in small viewports (see the screenshot below). They may need to be even smaller for small viewports and/or wrap around. You can use the tailwindcss modifiers to adjust styling based on viewport size.
> ![image](https://github.com/user-attachments/assets/c70f2561-c486-40fa-af06-8666b6a1b5af)
> 
> - The "copy" icon seems to scale down in smaller viewports, while the others don't. All of them should be styled consistently.
> 
> - When message A is selected, then message B is clicked, lets have it only deselect Message A, and require the user to click message B again to select it. I think that UX feels more similar to other apps.
> 
> - If you import a new js package, please add it to package.json. It looks like `@tauri-apps/plugin-dialog` was not added. (I went ahead and added it and pushed to this branch)


Resolved:

1. Removed the "cancel" button; clicking outside deselects the message.


2. Ensured ensured proper scaling/wrapping on small viewports with TailwindCSS modifiers.


3. Adjusted behavior to deselect Message A when Message B is clicked, requiring a second click to select Message B.


4. Ensured all new JS packages are added to package.json. Added @tauri-apps/plugin-dialog.

mattyg commented on package.json:
```diff
@@ -31,6 +31,7 @@
     "@tsconfig/svelte": "^5.0.4",
     "@types/dompurify": "^3.0.5",
     "@types/lodash-es": "^4.17.12",
+    "@types/node": "^22.10.1",
```
what is this needed for?

MasterK0927 replied:
Type definitions for nodejs was missing for typescript. Don't know if it's specific for my system. 

```
let configTimeout : NodeJS.Timeout
 let agentTimeout : NodeJS.Timeout
 let messageTimeout : NodeJS.Timeout

```

mattyg replied:
ah gotcha - thanks for adding that!

mattyg commented:
@MasterK0927 It looks like we can use tauri-plugin-fs for writing the blob to a file at the path the user chose in the save dialog. Here's the setup docs: https://tauri.app/plugin/file-system/

Do you want to give that a try?

MasterK0927 commented:
> @MasterK0927 It looks like we can use tauri-plugin-fs for writing the blob to a file at the path the user chose in the save dialog. Here's the setup docs: https://tauri.app/plugin/file-system/
> 
> Do you want to give that a try?

I already tried it, there was the same issue with this too.

mattyg committed:
Merge branch 'develop-0.3' into feat/select-message

mattyg committed:
feat(ui): saving images working on desktop & android

mattyg committed:
revert: accidentally committed, unrelated

mattyg commented on src-tauri/gen/schemas/acl-manifests.json:
just FYI these gen/schemas files are all auto-generated artifacts from tauri

mattyg committed:
fix(ui): convert image url to blob during download step

mattyg commented:
> I already tried it, there was the same issue with this too.

I got image download working using tauri-plugin-dialog and tauri-plugin-fs. I pushed the changes to this branch -- please review the diff to understand how it is implemented: https://github.com/holochain-apps/volla-messages/pull/292/files/3f1c6d533a9df2317707b98f5b5ac0ff5a405577..bf0844a3ff47dafc73003cdb82a112bc81267a85

MasterK0927 commented:
> > I already tried it, there was the same issue with this too.
> 
> @MasterK0927  I got image download working on android using tauri-plugin-dialog and tauri-plugin-fs. I pushed the changes to this branch -- please review the diff to understand how it is implemented: https://github.com/holochain-apps/volla-messages/pull/292/files/3f1c6d533a9df2317707b98f5b5ac0ff5a405577..bf0844a3ff47dafc73003cdb82a112bc81267a85


I tried the same approach earlier but it didn't worked for me. 

Can you please confirm, changes made in this file is from your side `src-tauri/capabilities/main.json` or it's auto generated by tauri?

mattyg commented:
> Can you please confirm, changes made in this file is from your side `src-tauri/capabilities/main.json` or it's auto generated by tauri?

Yes I made those changes -- we have to specify additional permissions of the tauri runtime to use plugins.

MasterK0927 commented:
> > Can you please confirm, changes made in this file is from your side `src-tauri/capabilities/main.json` or it's auto generated by tauri?
> 
> Yes I made those changes -- we have to specify additional permissions of the tauri runtime to use plugins.

Oh! Then it must be the cause, it was not working for me.

mattyg commented:
> Oh! Then it must be the cause, it was not working for me.

When debugging its helpful to look at the console logs in the webview to see errors. You can do so by:
- connect the android device to your computer
- open the app on the android device
- open chrome or chromium on your computer
- navigate to `chrome://inspect/devices#devices` in chrome
- under Remove Targets you should see the webview on the android device, and you can use the dev tools to view console logs

Or alternatively you can use `adb logcat` to see tauri errors.

mattyg commented:
@MasterK0927 This is just waiting on fixing "Download icon used had different color." before I re-review it, right?

MasterK0927 commented:
> @MasterK0927 This is just waiting on fixing "Download icon used had different color." before I re-review it, right?
> 

Yeah, that's the only issue left, I will fix it now.

MasterK0927 committed:
fixed the icon color for download

mattyg committed:
Merge branch 'develop-0.3' into feat/select-message

mattyg reviewed (changes requested):
- Now that #295 has merged, lets include success & error toasts for copy / download as needed
- ~After downloading the photo(s) the message should be de-selected~ actually looks like this is working
- It still looks like the download icon is a bit smaller than the copy icon in larger viewports (it has more padding within the svg I believe). I'm fine keeping it as is for now but just a heads up.
![image](https://github.com/user-attachments/assets/596b6106-d35a-42e4-a140-c5197e10ad2a)
- On macOS when a message contains multiple images, only the first image is downloaded. We may need to wait for the first image download to complete before triggering the next

MasterK0927 committed:
adds success and error toasts for downlad and copy buttons upon selecting message

MasterK0927 committed:
adds sequential downloading for multiple images

MasterK0927 commented:
> * Now that [feat/success and error toasts #295](https://github.com/holochain-apps/volla-messages/pull/295) has merged, lets include success & error toasts for copy / download as needed
> 
>     * ~After downloading the photo(s) the message should be de-selected~ actually looks like this is working
> 
>     * It still looks like the download icon is a bit smaller than the copy icon in larger viewports (it has more padding within the svg I believe). I'm fine keeping it as is for now but just a heads up.
>       ![image](https://private-user-images.githubusercontent.com/159270/394927344-596b6106-d35a-42e4-a140-c5197e10ad2a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM5OTQ1NzksIm5iZiI6MTczMzk5NDI3OSwicGF0aCI6Ii8xNTkyNzAvMzk0OTI3MzQ0LTU5NmI2MTA2LWQzNWEtNDJlNC1hMTQwLWM1MTk3ZTEwYWQyYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxMlQwOTA0MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZTVkODA3NDkxM2NjNTA2NjAwN2UzZjY1M2I0MmQzZjAxOTNjZGJjMzIwY2U1NmZlOGEwODhmYzRmNzkxMTg5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FitkBy6ukATtmDcE9vqR-lMjI5QHECmGCrVyBdpS3io)
> 
>     * On macOS when a message contains multiple images, only the first image is downloaded. We may need to wait for the first image download to complete before triggering the next

- Added sequential downloading for threaded (multiple) images.
- Added toasts for copy and download buttons upon selecting messages.
